### ISSUE TITLE
[QUERY] Client side encryption using `azure-storage-blob-cryptography` library

### ISSUE THREAD

**Issue by SukruthKS** [Azure/azure-sdk-for-java#6536]
**Query/Question**
We are currently using client side encryption functionality provided in v8 SDK. I'm upgrading our application to use v12 and I have a few questions/issues:

- Passing a key to `EncryptedBlobClient` requires implementing `AsyncKeyEncryptionKey` interface. Are there any existing implementations I can use? With v8, we were using `SymmetricKey` class provided in `com.microsoft.azure.keyvault.cryptography`.

- `EncryptedBlobClient` does not support uploading a blob from a stream like in `BlockBlobClient`. It only supports uploading from a file. Is there a way to do it using a stream? I'd prefer not to create files on disk just to do this.

- `EncryptedBlobClient` cannot be created from `BlobContainerClient` like other clients (e.g., `BlockBlobClient`). In our application, we'd initialize `BlobServiceClient` once and derive other clients like container or blob depending on the access needs. Since `EncryptedBlobClient` requires full standalone construction, I either need to cache credentials (all supported types as any of them could be used) or always create `EncryptedBlobClient` when creating `BlobServiceClient` even if client side encryption is not going to be used. 

- Client side encryption feature was seamlessly integrated in v8 compared to v12. Depending on whether a blob needs to be encrypted, I'd have to use two different clients in v12 whereas in v8 I just have to set `BlobEncryptionPolicy` in upload options.

***Why is this not a Bug or a feature Request?***
NA

**Setup (please complete the following information if applicable):**
 - OS: Linux
 - IDE : IntelliJ
 - Version of the Library used: 12.0.0
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by rickle-msft** [comment]
Hi, @SukruthKS. Responding to your points in order:

- The keyvault.keys.cryptography package has a KeyEncryptionKeyBuilder that you can use to generate implementations of the type to pass to the encrypted blob builder. The builder requests an async key encryption key, but only because we use one internally. I don't think you should really have to deal with any async objects because of this interface, but please let me know if I have overlooked something. 
- I am planning on doing some refactor work to BlobOutputStream that I think should enable this scenario. We are aware of the need and are tracking it, but it is not currently supported. You can track this at #5715
-  We'd love to get your feedback on this one before we commit to anything too strongly. What I think might work is if we have the builders accept another client and populate the builder based on those fields (I sound like a broken record). In that case, if you decide you need an encrypted blob client, you can initialize the builder with the blob client you already have. Could that enable your current scenarios? We can't put a getEncrytpedBlobClient method on BlobContainerClient because that would force blobs to have a dependency on encryption, which we want to avoid, but I'm hoping the solution I proposed will be almost as convenient? 
- We made an intentional decision to split the packages because of the dependency, so to a certain degree it will never be quite as seamless as v8, but if we fix everything else listed here, it should hopefully approach what we had in v8.

**Comment by SukruthKS** [comment]
Thanks @rickle-msft !

> The keyvault.keys.cryptography package has a KeyEncryptionKeyBuilder that you can use to generate implementations of the type to pass to the encrypted blob builder. The builder requests an async key encryption key, but only because we use one internally. I don't think you should really have to deal with any async objects because of this interface, but please let me know if I have overlooked something.

I looked at [KeyEncryptionKeyClientBuilder](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java) and this requires having a key in key vault. Whereas [SymmetricKey](https://github.com/Azure/azure-keyvault-java/blob/master/azure-keyvault-cryptography/src/main/java/com/microsoft/azure/keyvault/cryptography/SymmetricKey.java) implementation we are using with v8, encryption is performed locally on the machine and we provide the key. Is there something similar for the v12 interface?

> We'd love to get your feedback on this one before we commit to anything too strongly. What I think might work is if we have the builders accept another client and populate the builder based on those fields (I sound like a broken record). In that case, if you decide you need an encrypted blob client, you can initialize the builder with the blob client you already have. Could that enable your current scenarios? We can't put a getEncrytpedBlobClient method on BlobContainerClient because that would force blobs to have a dependency on encryption, which we want to avoid, but I'm hoping the solution I proposed will be almost as convenient?

Yes, that should work!

**Comment by rickle-msft** [comment]
@SukruthKS 
#6558 can track the work for the builders. 

> I looked at KeyEncryptionKeyClientBuilder and this requires having a key in key vault. Whereas SymmetricKey implementation we are using with v8, encryption is performed locally on the machine and we provide the key. Is there something similar for the v12 interface?

I have checked with the KeyVault team, and there is not currently an implementation that supports this. @g2vinay works on the key vault sdk and can answer more questions on that. 

At this point it seems to me like the best course is to open an issue against key vault for this key implementation and close this query thread if you think all the questions are appropriately addressed and tracked with other issues.

**Comment by SukruthKS** [comment]
Created a feature request for it - #6569

**Comment by SukruthKS** [comment]
@rickle-msft can you please provide me a sample on how to use `EncryptedBlobClient`? I'm trying to upload a blob using this client and here's the code I used. 

```
EncryptedBlobClient encryptedBlobClient = new EncryptedBlobClientBuilder()
    .endpoint(serviceClient.getAccountUrl())
    .pipeline(serviceClient.getHttpPipeline())
    .containerName(containerName)
    .blobName(blobName)
    .key(akek, KeyWrapAlgorithm.A256KW.toString())
    .buildEncryptedBlobClient();

encryptedBlobClient.upload(inputStream, content.length(), true);
```

The `upload() ` method is creating a blob in plaintext and does not seem to use the key I provide. Am I not adding something to indicate that the blob should be encrypted?

**Comment by rickle-msft** [comment]
Hi, @SukruthKS. Pipelines are immutable, so as it stands now, when you specify a pipeline, we ignore any other configuration options set on the builder in favor of using the given pipeline. Thus, the key you are passing in later in the chain is getting ignored, and since the pipeline from your serviceClient doesn't have an encryption policy, there ends up being no encryption policy. 

It seems maybe we've been to restrictive in the builder behavior and may need to rethink it a bit? Can you give us some more details on your workflow so we can be sure we capture the scenario in our fix?
This seems related to the work we'll need to do for #6558

**Comment by SukruthKS** [comment]
Thanks for the quick reply!

In our workflow, we construct `BlobServiceClient` once (with appropriate credentials) for a logical application operation and of course, this `BlobServiceClient` instance is not directly exposed to the application. The application can then perform any type and number of storage operations including uploading some blobs with client-side encryption (using different keys) and some without. So, we need to be able to derive other specialized clients like `EncryptedBlobClient` from `BlobServiceClient` for individual operations.

**Comment by rickle-msft** [comment]
Ok. We will get working on that next week and hopefully have something out for you soon.

**Comment by SukruthKS** [comment]
Thanks, Rick! 
I built the encryption client without passing an existing pipeline like below, but the blob is still not getting encrypted. The encryption key I'm providing is a local symmetric key and constructed using the sample code provided by Vinay in https://github.com/Azure/azure-sdk-for-java/issues/6569

```
EncryptedBlobClient encryptedBlobClient = new EncryptedBlobClientBuilder()
    .endpoint(serviceClient.getAccountUrl())
    .sasToken("<SAS token>")
    .containerName(containerName)
    .blobName(blobName)
    .key(akek, KeyWrapAlgorithm.A256KW.toString())
    .buildEncryptedBlobClient();

encryptedBlobClient.upload(inputStream, content.length(), true);
```

**Comment by SukruthKS** [comment]
@rickle-msft following Vinay's suggestion in https://github.com/Azure/azure-sdk-for-java/issues/6569, I used a key Id that is parsable by the library. The issue I'm seeing is that `EncryptedBlobClient` encrypts the blob if `uploadFromFile()` API is used, but it doesn't if `upload(InputStream, long, boolean)` is used.

**Comment by rickle-msft** [comment]
@SukruthKS Thanks for the update. I just came to the same conclusion a few minutes ago and was discussing with a teammate when you posted this. We recently added this method to BlobClient, and it seems we didn't properly override it in EncryptedBlobClient. For safety, we'll put out a hotfix tomorrow that either throws an exception or includes proper behavior for this method (depending on what we can realistically do). If we opt to throw in the next patch release tomorrow, we'll follow up shortly after to add actual support for this. 

In the meantime, BlobOutputStream also works as expected if that's more convenient than using uploadFromFile
